### PR TITLE
mean, median and max message lag metrics

### PR DESCRIPTION
**Pull request by robfletcher**

### CODE REVIEW

**Comment by robfletcher** [comment]
I figure we can overlay these on a single Atlas graph and it could provide a useful signal of Orca health.

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt [hunk 1]
```diff
@@ -96,8 +101,18 @@ class AtlasQueueMonitor
         .toMillis()
         .toDouble()
     })
-    registry.gauge("queue.message.lag", this, {
-      it.averageMessageLag
+    registry.gauge("queue.mean.lag", this, {
```
Comment (robfletcher) [inline orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt]: gauge for mean lag

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt [hunk 2]
```diff
@@ -96,8 +101,18 @@ class AtlasQueueMonitor
...
+      it.meanMessageLag
+        .toMillis()
+        .toDouble()
+    })
+    registry.gauge("queue.median.lag", this, {
```
Comment (robfletcher) [inline orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt]: gauge for median lag

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt [hunk 3]
```diff
@@ -96,8 +101,18 @@ class AtlasQueueMonitor
...
+      it.medianMessageLag
+        .toMillis()
+        .toDouble()
+    })
+    registry.gauge("queue.max.lag", this, {
```
Comment (robfletcher) [inline orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt]: gauge for max lag

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt [hunk 4]
```diff
@@ -121,15 +136,25 @@ class AtlasQueueMonitor
...
-        .average()
-        .let { Duration.ofMillis(it.toLong()) }
-      clear()
-      return avg
-    }
-  private val _messageLags = mutableListOf<Duration>()
+  val meanMessageLag: Duration
+    get() = _meanMessageLag
+      .getAndSet(emptyList())
+      .average()
```
Comment (robfletcher) [inline orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt]: get an average of the lag and clears the register

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt [hunk 5]
```diff
@@ -121,15 +136,25 @@ class AtlasQueueMonitor
...
+      .let { Duration.ofMillis(it.toLong()) }
+  private val _meanMessageLag = AtomicReference<List<Long>>(emptyList())
+
+  val medianMessageLag: Duration
+    get() = _medianMessageLag
+      .getAndSet(emptyList())
+      .median()
```
Comment (robfletcher) [inline orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt]: gets the median lag and clears the register

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt [hunk 6]
```diff
@@ -121,15 +136,25 @@ class AtlasQueueMonitor
...
+      .let { Duration.ofMillis(it.toLong()) }
+  private val _medianMessageLag = AtomicReference<List<Long>>(emptyList())
+
+  val maxMessageLag: Duration
+    get() = _maxMessageLag
+      .getAndSet(0)
```
Comment (robfletcher) [inline orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt]: gets the max lag and clears the register

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt [hunk 7]
```diff
@@ -189,3 +214,10 @@ class AtlasQueueMonitor
   private val MessageNotFound.counter: Counter
     get() = registry.counter("queue.message.notfound")
 }
+
+private fun List<Long>.median(): Double =
+  when {
+    isEmpty() -> Double.NaN
+    size % 2 == 1 -> sorted()[size / 2].toDouble()
+    else -> sorted().subList((size / 2) - 1, (size / 2) + 1).average()
+  }
```
Comment (robfletcher) [inline orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt]: My implementation of getting the median of a list of longs

File: orca-queue/src/test/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitorTest.kt [hunk 8]
```diff
@@ -92,28 +94,54 @@ object AtlasQueueMonitorTest : SubjectSpek<AtlasQueueMonitor>({
...
-      it("averages the lag time") {
-        subject.averageMessageLag shouldEqual lag.map { it.toMillis() }.average().let { Duration.ofMillis(it.toLong()) }
+    describe("message lag metrics") {
+      given("no messages have been processed") {
+        afterGroup(::resetMocks)
+
+        it("reports zero lag time") {
+          assertThat(subject.meanMessageLag).isEqualTo(ZERO)
+          assertThat(subject.medianMessageLag).isEqualTo(ZERO)
+          assertThat(subject.maxMessageLag).isEqualTo(ZERO)
```
Comment (robfletcher) [inline orca-queue/src/test/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitorTest.kt]: all the gauges should be zero if nothing has happened

File: orca-queue/src/test/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitorTest.kt [hunk 9]
```diff
@@ -92,28 +94,54 @@ object AtlasQueueMonitorTest : SubjectSpek<AtlasQueueMonitor>({
...
+        }
+
+        on("receiving a ${events.first().javaClass.simpleName} event") {
+          events.forEach(subject::onQueueEvent)
+        }
+
+        it("averages the lag time") {
+          assertThat(subject.meanMessageLag).isEqualTo(lag.average())
+          // after reading the mean should reset
+          assertThat(subject.meanMessageLag).isEqualTo(ZERO)
```
Comment (robfletcher) [inline orca-queue/src/test/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitorTest.kt]: mean lag is reported and reset

File: orca-queue/src/test/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitorTest.kt [hunk 10]
```diff
@@ -92,28 +94,54 @@ object AtlasQueueMonitorTest : SubjectSpek<AtlasQueueMonitor>({
...
+        }
+
+        it("records the median lag time") {
+          assertThat(subject.medianMessageLag).isEqualTo(lag.sorted().toList()[1])
+          // after reading the mean should reset
+          assertThat(subject.medianMessageLag).isEqualTo(ZERO)
```
Comment (robfletcher) [inline orca-queue/src/test/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitorTest.kt]: median lag is reported and reset

File: orca-queue/src/test/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitorTest.kt [hunk 11]
```diff
@@ -92,28 +94,54 @@ object AtlasQueueMonitorTest : SubjectSpek<AtlasQueueMonitor>({
...
+        }
+
+        it("records the max lag time") {
+          assertThat(subject.maxMessageLag).isEqualTo(lag.max())
+          // after reading the max should reset
+          assertThat(subject.maxMessageLag).isEqualTo(ZERO)
```
Comment (robfletcher) [inline orca-queue/src/test/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitorTest.kt]: max lag is reported and reset

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt [hunk 12]
```diff
@@ -49,8 +50,12 @@ class AtlasQueueMonitor
   @EventListener
   fun onQueueEvent(event: QueueEvent) {
     when (event) {
-      is QueuePolled -> _lastQueuePoll.set(clock.instant())
-      is MessageProcessing -> _messageLags.add(event.lag)
+      QueuePolled -> _lastQueuePoll.set(clock.instant())
+      is MessageProcessing -> {
+        _meanMessageLag.updateAndGet { it + event.lag.toMillis() }
```
Comment (cfieber) [inline orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/metrics/AtlasQueueMonitor.kt]: I think updateAndGet here has a concurrency issue if `it + event.lag.toMillis()` isn't side-effect free and is operating on the underlying list object. Could be I'm wrong if the kotlin `+` is returning a new list object with the result appended..

**Comment by cfieber** [comment]
I think it might be more appropriate to just record a spectator timer for message lag and do the aggregation on the graph side?

**Pull request merged**